### PR TITLE
add in hand optimized bril, add support for fn args

### PR DESCRIPTION
**Pull request by egg-fnatic**
I've added support for functions with arguments, although I'm not entirely certain it is sound (looking at you ReturnVal).

So far the tests are just making sure that everything passes through okay, but I've left comments on the top of each test to show the optimized version of each for when we actually attempt to optimize otherwise we will constantly fail tests

### CODE REVIEW

**Comment by oflatt** [comment]
Great work-
Right now optimized version is in a comment
Could you add the optimized versions as bril code in another file? You could just add like a `.optimized.bril` extension.
For now they won't pass (the optimized version won't match the target exactly) so have a list of tests that shouldn't pass or something like that.

File: tests/reassociate_mul.bril [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# (x * (5 * (y * (6 * z)))) -> (x * (y * (z * 30)))
+@mul_reassoc (x: int, y: int, z: int): int {
```
Comment (oflatt) [inline tests/reassociate_mul.bril:2]: might be a problem if main doesn't use this function- might get dead code eliminated in the future

File: src/lib.rs [hunk 2]
```diff
@@ -61,14 +60,17 @@ impl Default for Optimizer {
 }
 
 impl Optimizer {
-    pub fn parse_and_optimize(&mut self, program: &str) -> Result<Program, EggCCError> {
+    pub fn parse_and_optimize(
+        &mut self,
+        program: &str,
+    ) -> Result<Program, Box<dyn std::error::Error>> {
```
Comment (oflatt) [inline src/lib.rs:66]: I don't think this should be a `std::error::Error`. The benefit of the `EggCCError` is we control the output- we want to output good error messages that have meaning to the user of eggcc, not random rust errors

File: tests/reassociate_mul.bril (hunk 1 above)
Comment (egg-fnatic) [reply tests/reassociate_mul.bril:2]: We might similarly have an issue if we do call the function and an inliner determines it to be inlineable :)

I lean towards not doing deadcode on functions for now, unless we're pushing for codegen size as a metric of success which I don't think we should/are.

Comment (oflatt) [reply tests/reassociate_mul.bril:2]: Good point, lets just deal with this if it becomes a problem

**Comment by egg-fnatic** [comment]
@oflatt We should probably just add in the hand-optimized bril and kill the rest. I'll see if I can get this done soonish as it isn't a large priority.

The test cases that I added for hand-optimized bril will help us from getting any sort of blowup in associativity/commutativity which will be nice to have

**Comment by oflatt** [comment]
sounds good to me

**Comment by oflatt** [comment]
@ryan-berger I think these benchmarks are still valuable, esp. because I implemented reassociation for addition
Would you open a PR for these benchmarks? You can even take a crack at optimizing them if you want

**Comment by oflatt** [comment]
closing since this is quite out of date

**Pull request closed without merging**